rpc/wallet: add simulaterawtransaction RPC

kallewoof opened the pull request:
(note: this was originally titled "add analyzerawtransaction RPC")

This command iterates over the inputs and outputs of the given transactions, and tallies up the balance change for the given wallet. This can be useful e.g. when verifying that a coin join like transaction doesn't contain unexpected inputs that the wallet will then sign for unintentionally.

I originally proposed this to Elements (https://github.com/ElementsProject/elements/pull/1016) and it was suggested that I propose this upstream.

There is an alternative #22776 to instead add this info to `getbalances` when providing an optional transaction as argument.

kallewoof committed:
rpc/wallet: add simulaterawtransaction RPC

This command iterates over the inputs and outputs of the given transactions, and tallies up the balance change for the given wallet. This can be useful e.g. when verifying that a coin join like transaction doesn't contain unexpected inputs that the wallet will then sign for unintentionally.

fanquake commented on test/functional/wallet_analyzerawtx.py:
```diff
@@ -0,0 +1,64 @@
+        assert_approx(w0.analyzerawtransaction(tx), -(5.0 + bitcoin_fee))
+
+        # node1 sees same as before
+        analysis = w1.analyzerawtransaction(tx)
```
`test/functional/wallet_analyzerawtx.py:60:9: F841 local variable 'analysis' is assigned to but never used`

kallewoof replied:
Thanks, fixed.

meshcollider commented:
Concept ACK, but I don't think the RPC name represents the functionality. Analyze implies some sort of in-depth breakdown of information. Perhaps `nettransactionbalance` or something to represent the balance-specific use?

prayank23 commented:
> but I don't think the RPC name represents the functionality. Analyze implies some sort of in-depth breakdown of information. 

Agree. We already have RPC with similar name: `analyzepsbt` which returns lot of things that would be helpful for someone using PSBT. Although I am still confused between `decodepsbt` and `analyzepsbt`.

> Perhaps nettransactionbalance or something to represent the balance-specific use?

Or maybe add it in the results of any existing RPC

kallewoof commented:
I have no strong feelings about keeping `analyze` but I think `<verb>rawtransaction` is a good name, where `<verb>` somehow indicates looking at each input and output and determining whether they belong to the wallet, and if so to add/subtract the appropriate amount.

prayank23 commented:
> adding to results of existing RPC sounds good to me too, but not sure which that would be.

Can we add this to [`getbalance`](https://bitcoincore.org/en/doc/0.21.0/rpc/wallet/getbalance/)? Will need one argument `include_tx` so the command would look like this:

```
$bitcoin-cli -named getbalance include_tx=hex

Result:
n    (numeric) The total amount in BTC received for this wallet.
n    (numeric) The wallet balance change including tx(negative means decrease).
```

jonatack reviewed (commented):
Concept ACK. Maybe add to decoderawtransaction? (Ignore me if that's dumb, and feel free to ignore the more detailed review comments below until the direction is set.)

kallewoof commented:
@prayank23 

> Can we add this to [`getbalance`](https://bitcoincore.org/en/doc/0.21.0/rpc/wallet/getbalance/)?

Adding to `getbalance` sounds reasonable to me.

@jonatack 

> Concept ACK. Maybe add to decoderawtransaction? (Ignore me if that's dumb, and feel free to ignore the more detailed review comments below until the direction is set.)

This requires the wallet so `decoderawtransaction` is out, I'm afraid. :)

jonatack commented:
(getbalances is the replacement for getbalance).

kristapsk commented:
I think it would be better to return JSON instead of numeric, that would allow adding possible other analysis in the future without breaking backwards compatibility.

kiminuo commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4567,6 +4567,59 @@ static RPCHelpMan upgradewallet()
+
+    // Decode the transaction
+    CMutableTransaction mtx;
+    if (!DecodeHexTx(mtx, request.params[0].get_str(), true, true)) {
```
Nit: Consider please:
```suggestion
    if (!DecodeHexTx(mtx, request.params[0].get_str(), /* try_no_witness */ true, /* try_witness */ true)) {
```

kiminuo commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4567,6 +4567,59 @@ static RPCHelpMan upgradewallet()
+        },
+    [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
+    std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(request);
```
Nit: People typically place `const`s before types in this codebase, AFAIK. If I'm wrong, please correct me.

kallewoof replied:
Looking around in the code, the way I wrote it seems to be the way it's written elsewhere, so keeping as is.

kiminuo replied:
Two lines above the `const` modifier is before type. Anyway, it's not important so I don't want to waste your time on this any further.

kallewoof replied:
That would actually make the shared pointer a constant, which points to a non-constant CWallet instance.

See e.g. https://stackoverflow.com/questions/7526152/easy-rule-to-read-complicated-const-declarations

kallewoof replied:
Also try modifying my PR to this, and try compiling:
```C++
    const std::shared_ptr<CWallet> wallet = GetWalletForJSONRPCRequest(request);
    if (!wallet) return NullUniValue;
    wallet->ResendWalletTransactions(); // <- this is not const
```
(it works, even though ResendWalletTransactions() is NOT const-declared)

kiminuo replied:
Thanks for the clarification!

kallewoof replied:
Actually, I'm mistaken, and the two versions are identical. (ResendWalletTransactions() compiles for both cases.)
I actually think this is a mistake, and could just as well be
```C++
std::shared_ptr<CWallet> wallet = GetWalletForJSONRPCRequest(request);
```
but I'm just going to keep the const where it is, as that's the case for other RPC calls. Sorry for the misinformation!

kallewoof replied:
I opened a separate pull request on this subject. Thanks for the nudge. https://github.com/bitcoin/bitcoin/pull/22787

kiminuo commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4567,6 +4567,59 @@ static RPCHelpMan upgradewallet()
+    CAmount changes = 0;
+
+    // Fetch debit; we are *spending* these; if the transaction is signed and broadcast, we will lose everything in these
+    for (size_t i = 0; i < mtx.vin.size(); ++i) {
```
Would it be more idiomatic to use `for (CTxIn& txin : mtx.vin) {`?

kallewoof replied:
Done.

kiminuo commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4567,6 +4567,59 @@ static RPCHelpMan upgradewallet()
+    if (!wallet) return NullUniValue;
+    CWallet* const pwallet = wallet.get();
+
+    RPCTypeCheck(request.params, {UniValue::VSTR}, true);
```
```suggestion
    RPCTypeCheck(request.params, {UniValue::VSTR}, /* fAllowNull */ true);
```

kiminuo replied:
If you allow `null`, then L4598 will lead to a crash, would it not?

kallewoof replied:
The RPCHelpMan framework will not execute the code, since the first param is non-optional.

kiminuo replied:
Would `RPCTypeCheck(request.params, {UniValue::VSTR})` work too then?

kallewoof commented:
Thanks for all the feedback. I'm rewriting this as a separate pull request that adds the feature to `getbalances`. I'll potentially keep updating this PR as well, in case somebody prefers this variant.

kallewoof commented:
Opened alternative #22776 and updated this PR.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4685,6 +4738,7 @@ static const CRPCCommand commands[] =
     { "wallet",             &unloadwallet,                   },
     { "wallet",             &upgradewallet,                  },
     { "wallet",             &walletcreatefundedpsbt,         },
+    { "wallet",             &analyzerawtransaction,          },
```
Thanks, addressed all. See also alt PR.

kristapsk reviewed (approved):
ACK f19676bf02cb9cc279e2439c53b1f54ff53328fa

codedfile reviewed (commented):
?

kristapsk reviewed (approved):
re-ACK c3628daf58bbe8d745cef08db8cee5fffd95167e

kiminuo commented on test/functional/wallet_analyzerawtx.py:
```diff
@@ -0,0 +1,64 @@
+        # w0 funds transaction; it should now see a decrease in (tx fee and payment), and w1 should see the same as above
+        funding = w0.fundrawtransaction(tx)
+        tx = funding["hex"]
+        bitcoin_fee = float(funding["fee"])
```
Would it be better to use `Decimal` here?

kallewoof replied:
Could, but would be another import just for this one place, and the code should be fine (i.e. no rounding errors) due to the `assert_approx`.

meshcollider commented:
@kallewoof do you mind adding mention of #22776 to the OP so people looking at this are clear of the alternative

kallewoof commented:
> @kallewoof do you mind adding mention of #22776 to the OP so people looking at this are clear of the alternative

Good point, done.

kallewoof commented:
Apologies, I switched to a constant CWallet so had to re-push.

laanwj commented:
Concept ACK, interesting feature

achow101 commented on test/functional/wallet_simulaterawtx.py:
```diff
@@ -0,0 +1,88 @@
+class SimulateTxTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 2
```
In 4d6c1a01532fa189c44714fae31916d758f3a574 "rpc/wallet: add simulaterawtransaction RPC"

I would prefer using one node with 2 wallets. Having multiple nodes, with one wallet on each, is how we used to do things before multiwallet. Given that the test already makes testing wallets (one on each node), it's unnecessary to have multiple nodes.

kallewoof replied:
That is obviously the right way to do it. Thanks, fixed.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4567,6 +4567,76 @@ static RPCHelpMan upgradewallet()
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Transaction(s) are spending the same output more than once");
+            }
+            spent.insert(txin.prevout);
+            changes -= pwallet->GetDebit(txin, ISMINE_SPENDABLE);
```
Sounds good - adding an `include_watchonly` option.

kallewoof commented:
* added `include_watchonly` option
* updated tests to use one node with two wallets
* added `include_watchonly` tests

prayank23 commented:
Approach NACK

I prefer https://github.com/bitcoin/bitcoin/pull/22776

Reasons: 

1. No need to create another RPC and remember another RPC.
2. Other PR keeps things simple. This is basically returning balances but based on some conditions. It makes sense to add conditions (transactions in this case) as an argument in `getbalances`.
3. Same approach can be followed in future for other RPCs if required. Example: What will be output for `getmempoolinfo` if a list of transactions are broadcasted: `getmempoolinfo "[Tx1, Tx2.. Txn]"`

kallewoof commented:
FWIW, @apoelstra suggests this is modified to take PSBT(s) instead of raw transactions, stating that the latter is more or less made obsolete by the former, and in multi-party transactions people will most likely be using PSBT.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4660,6 +4660,95 @@ static RPCHelpMan upgradewallet()
+    RPCTypeCheck(request.params, {UniValue::VARR, UniValue::VOBJ}, true);
+
+    LOCK(pwallet->cs_wallet);
+    EnsureWalletIsUnlocked(*pwallet);
```
In 40944a573a21c225662ce0c8ba7f64a8427ca22d "rpc/wallet: add simulaterawtransaction RPC"

This is unnecessary since no key access is needed.

kallewoof replied:
Do the vin/vout checks not access keys to determine if the wallet owns them?

achow101 replied:
It only uses pubkeys.

kallewoof replied:
OK, you're talking about the `EnsureWalletIsUnlocked` line. I thought you were also talking about L4698. Fixed.

MarcoFalke commented on test/functional/wallet_simulaterawtx.py:
```diff
@@ -0,0 +1,97 @@
+        w0 = node.get_wallet_rpc('w0')
+        w1 = node.get_wallet_rpc('w1')
+
+        node.generatetoaddress(nblocks=COINBASE_MATURITY + 1, address=w0.getnewaddress())
```
Needs to use self.gener... after a rebase

kallewoof replied:
Thanks, updated.

kallewoof commented:
Rebased. I don't even *see* an error in the two failing tests.

laanwj commented:
For what it's worth I prefer this as a separate RPC over #22776. Trying to squash this into `getbalances` seems a bit weird to me and I agree with the reasoning in https://github.com/bitcoin/bitcoin/pull/22751#issuecomment-905950000 that this can do more, this way.

laanwj commented on test/functional/wallet_simulaterawtx.py:
```diff
@@ -0,0 +1,97 @@
+        # w0 funds transaction; it should now see a decrease in (tx fee and payment), and w1 should see the same as above
+        funding = w0.fundrawtransaction(tx1)
+        tx1 = funding["hex"]
+        bitcoin_fee = float(funding["fee"])
```
Would be better to use `Decimal` instead of float for monetary amounts.

kallewoof committed:
test: add support for Decimal to assert_approx

kallewoof commented:
Switched to using Decimal. Also added support for mixing floats and Decimals to `assert_approx` as a separate commit.

kallewoof commented:
So if I understand the CI error right, it's no longer supported to import pubkeys to wallets that have privkeys (but nodes can?) so I should revert back to two nodes with separate wallets I assume? Sounds like a regression, but if it's the way it is, I'm fine doing that.

MarcoFalke commented on test/functional/wallet_simulaterawtx.py:
```diff
@@ -0,0 +1,99 @@
+        address2 = w1.getnewaddress()
+
+        # Make address1 watch-only in w0
+        w0.importpubkey(pubkey=w1.getaddressinfo(address1)["pubkey"])
```
`importpubkey` does not exist for descriptor wallets (sqlite).

`                                   test_framework.authproxy.JSONRPCException: Cannot import descriptor without private keys to a wallet with private keys enabled (-4)`

kallewoof replied:
Maybe I should rethink the usage of importpubkey completely. It seems to be not very well supported anymore.

achow101 commented:
> So if I understand the CI error right, it's no longer supported to import pubkeys to wallets that have privkeys (but nodes can?) so I should revert back to two nodes with separate wallets I assume? Sounds like a regression, but if it's the way it is, I'm fine doing that.

Yes. Part of the descriptor wallet changes is to have a clear separation of what is watch-only and what is spendable. This is achieved by having watch-only wallets which do not have private keys, and disallowing descriptor wallets with private keys to include watch-only things. So you should avoid continuing to use the legacy wallet behavior of mixing spendable and watch-only things in one wallet.

achow101 commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -582,6 +582,94 @@ static RPCHelpMan upgradewallet()
+{
+    return RPCHelpMan{"simulaterawtransaction",
+        "\nCalculate the balance change resulting in the signing and broadcasting of the given transaction(s).\n" +
+        HELP_REQUIRING_PASSPHRASE,
```
In ae8564c198a3f9c57cc896489afcb78d6b947b45 "rpc/wallet: add simulaterawtransaction RPC"

The wallet does not need to be unlocked, so this help text is not needed.

kallewoof replied:
Good point, removed.

achow101 commented on test/functional/wallet_simulaterawtx.py:
```diff
@@ -0,0 +1,99 @@
+        bitcoin_fee2 = Decimal(funding["fee"])
+
+        # w0 sees fees + 15 btc decrease, or fees + 10 btc decrease for watch only case
+        assert_approx(w0.simulaterawtransaction([tx1, tx2])["balance_change"], -(Decimal("5") + bitcoin_fee + Decimal("10") + bitcoin_fee2))
```
In ae8564c198a3f9c57cc896489afcb78d6b947b45 "rpc/wallet: add simulaterawtransaction RPC"

Since `tx1` is already sent, it will include inputs that are already spent. It seems like we should disallow transactions that spend already spent inputs, at least until we implement conflict checking in this RPC.

kallewoof replied:
Added fetch/check of coins for transactions and RPC asserts when coins are missing or spent.

JeremyRubin commented:
Approach wise, i think a better API might be `filterrawtxformine`, where it returns a list of inputs and outputs that are `IsMine`.

Although: I think this can already be done with a batch of gettxout and getaddressinfo. Why the need for a new API?

achow101 commented:
> Although: I think this can already be done with a batch of gettxout and getaddressinfo. Why the need for a new API?

For users who don't know the various ismine things and how the wallet actually calculates balances, it is preferable to have a built in RPC that does all that for them instead of hoping that what they think the wallet does is actually what it will do.

Additionally, I would like to see this RPC be expanded to be more than just balance calculation (in a future PR). I would like to have it check for conflicts, and that would require much more work if implemented externally.

JeremyRubin commented:
that seems like a recursive problem. A new wallet rpc will probably not do what we think a user should do?

Maybe it's more a matter of cleaning up / augmenting the behavior of things like ismine so that the behavior is clearer for API consumers?

JeremyRubin commented:
that said I don't think there's a terrible cost to a new RPC if it's useful, but I think we should have a concrete justification comparing it to the equivalent userland code. E.g., there have been cases where you can do it in userland but it is very very inefficient.

kallewoof commented:
> Yes. Part of the descriptor wallet changes is to have a clear separation of what is watch-only and what is spendable. This is achieved by having watch-only wallets which do not have private keys, and disallowing descriptor wallets with private keys to include watch-only things. So you should avoid continuing to use the legacy wallet behavior of mixing spendable and watch-only things in one wallet.

Understood! I've added a third watch-only wallet and moved the watch-only checks there.

achow101 commented on test/functional/wallet_simulaterawtx.py:
```diff
@@ -0,0 +1,97 @@
+        self.generate(node, 1, sync_fun=self.no_op) # Confirm tx to trigger error below
+        self.sync_all()
+
+        # w0 funds transaction 2; it should now see a decrease in (tx fee and payment), and w1 should see the same as above
+        funding = w0.fundrawtransaction(tx2)
+        tx2 = funding["hex"]
```
In aaa078857aa0bd4d881f781992dfc7c11aef0451 "rpc/wallet: add simulaterawtransaction RPC"

I think this section isn't doing anything now.

kallewoof replied:
Re-added tests for tx2 only.

achow101 commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -582,6 +582,109 @@ static RPCHelpMan upgradewallet()
+            if (spent.count(txin.prevout)) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Transaction(s) are spending the same output more than once");
+            }
+            if (!coins.count(txin.prevout)) {
```
In aaa078857aa0bd4d881f781992dfc7c11aef0451 "rpc/wallet: add simulaterawtransaction RPC"

If the UTXO points to one of the transactions in the array, then this would fail. I think it is possible someone would do this, especially with package relay on the horizon.

We could have `coins` outside of the `txs `loop, add the outputs of each tx to `coins`, and also remove coins as they are processed in this loop.

kallewoof replied:
To avoid things like tx1 spending tx2 and tx2 spending tx1, I think we may want to strictly require that transactions are in order. If we do, I think we can just iterate one at a time and just append new outputs as we go, without the extra round.

I updated the code to reflect this suggestion.

prayank23 commented:
> A new wallet rpc will probably not do what we think a user should do?

There is an alternative solution in https://github.com/bitcoin/bitcoin/pull/22776 which does not add any new RPC command

JeremyRubin commented:
@prayank23 my point was more that no code changes are required to use a constant number of RPC calls to answer this question. This technique might not have occurred to Kalle, so it was worth sharing.

Andrew's response ~was that "the already wallet does not do what you think it does" and that future changes might extend this RPC further.

I think adding an RPC makes more sense than changing and old one and both make less sense than not adding anything new unless there is a tangible issue with the mock algorithm i provided. A tangible issue could be "It's slow" or "race conditions between batches". I don't think we should have that issue here. Another compelling answer is "well it should eventually check for conflicts too, which we can't do well via just RPCs". In this case I would be curious to know if the check for conflict RPC should really be it's own RPC, and if we might benefit from thinking that through before pushing this.

I'm also worried that if there are defects in existing RPCs that make them unreliable somehow that we might fix those too.
I am curious -- for my benefit to learn -- why Andrew thinks the ismine field is not safe to use for this.

All that said, an RPC isn't giving anyone cancer so by all means, merge if it's useful.

kallewoof commented:
@JeremyRubin Primarily, my reason for wanting this is because I want a clear-cut way to say that the wallet changes I expect are happening. Yes I could manually write code to use getaddressinfo to do parts of this, but I *might get it wrong*. (Note that the command also includes new UTXO balance changes, which your pseudo-code does not, I think?)

JeremyRubin commented:
Yeah you should be able to add it onto the code i provided by just inspecting the inputs/outputs fetched with something like:

```rust
outs.iter().map(|v| v.1).sum()  - results.iter().map(|v| v.1).sum() 
```
-------------
i don't have a problem with this RPC existing, it does seem useful.

But I want to point out some other problematic behavior you might not have considered:

suppose I have a wallet that has an address X that ismine but has the script/descriptor: `and(older(2 years), key(mine))`

Suppose my wallet has outputs J under address A and K under address X that has matured.

then i ask my node to analyze a transaction that looks like:

Inputs: (J, 2.0) (K, 1.0), ....
Outputs: (A, 2.0), (X,  1.0), ....

where I was expecting the J -> A as my coinjoin, but K -> X was unexpected. Now my funds got automatically locked for another 2 years and this was not caught by this logic because the balance seemed fine.

------

IDK if we can have CSV scripts in descriptors today, but it's something to note that APIs like this are incredibly tricks.

I think the better version would be something like "I expect to care about input J and output A" and no other effect.

kallewoof commented:
> But I want to point out some other problematic behavior you might not have considered:

Those to me sound like good reasons to *have* a command like this, as it seems like a natural expansion to add feedback related to time locks in the response.

JeremyRubin commented:
i don't have a great answer for all this, it does seem to me that the best case is to make sure *the only* things matches are the ones you expect to be there, and that you *only* sign those.

Anything else could be a vuln.

achow101 commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -582,6 +582,124 @@ static RPCHelpMan upgradewallet()
+            spent.insert(txin.prevout);
+        }
+
+        for (size_t i = 0; i < txs.size(); ++i) {
```
In 84a745eed211b597a89f3bbc3da73b0ac2cb457d "rpc/wallet: add simulaterawtransaction RPC"

This loop doesn't make sense here. It shouldn't be necessary to iterate `txs` at the same time we are already iterating `txs`.

At best, this could be moved outside of the first `txs` loop as a preprocessing step to extract all of the UTXOs created, but I don't think that would be a good idea. That would allow for transactions that have circular dependencies and those would be invalid. It is fine to have an order requirement as that is what package validation requires.

kallewoof replied:
Yeah, this was clearly a mistake. I removed the loop and kept the inner block only.

achow101 commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -582,6 +582,124 @@ static RPCHelpMan upgradewallet()
+            // Populate new outputs
+            const auto hash = mtx.GetHash();
+            for (size_t i = 0; i < mtx.vout.size(); ++i) {
+                new_utxos[COutPoint(hash, i)] = (pwallet->IsMine(mtx.vout[i]) & filter) ? mtx.vout[i].nValue : 0;
```
In 84a745eed211b597a89f3bbc3da73b0ac2cb457d "rpc/wallet: add simulaterawtransaction RPC"

This can be moved into the loop below that iterates the outputs.

kallewoof replied:
Moved.

achow101 commented on test/functional/test_runner.py:
```diff
@@ -258,6 +258,7 @@
     'wallet_implicitsegwit.py --legacy-wallet',
     'rpc_named_arguments.py',
     'feature_startupnotify.py',
+    'wallet_simulaterawtx.py',
```
In 84a745eed211b597a89f3bbc3da73b0ac2cb457d "rpc/wallet: add simulaterawtransaction RPC"

Wallet functional tests should appear in test_runner.py twice, once with `--legacy-wallet` and once with `--descriptors`.

kallewoof replied:
Got it. Done.

kallewoof commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -583,6 +583,120 @@ static RPCHelpMan upgradewallet()
+        for (size_t i = 0; i < mtx.vout.size(); ++i) {
+            const auto& txout = mtx.vout[i];
+            bool is_mine = 0 < (pwallet->IsMine(txout) & filter);
+            changes += new_utxos[COutPoint(hash, i)] = is_mine ? txout.nValue : 0;
```
I hope this line isn't overly confusing.

achow101 commented:
ACK aff96844f7086458a414df9bb5cf9327a7266f2d

achow101 commented:
@meshcollider @jonatack @kristapsk @laanwj re-review this?

achow101 commented:
re-ACK bd520345f7a5076e5e1eb3877f6330a9e5816267

luke-jr commented on test/functional/test_framework/util.py:
```diff
@@ -28,6 +28,10 @@
+    if isinstance(v, Decimal) or isinstance(vexp, Decimal):
+        v=Decimal(v)
+        vexp=Decimal(vexp)
+        vspan=Decimal(vspan)
```
This results in `Decimal('0.000010000000000000000818030539140313095458623138256371021270751953125')`

Better to just insist `v` and `vexp` have compatible types, and default `vspan` sanely?

kallewoof replied:
The idea is to let the caller not worry too much about converting stuff to/from Decimal instances. Is there an issue resulting in `vspan` having some noise somewhere?

anibilthare commented:
Ack. Tested manually. Works fine.

I have also created a follow up [PR](https://github.com/bitcoin/bitcoin/pull/25621) which extends the functionality of this RPC.

achow101 commented:
ACK b557f53eac39b92bfc4922729e94087556de749a

furszy commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -590,6 +590,120 @@ static RPCHelpMan upgradewallet()
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Transaction hex string decoding failure.");
+        }
+
+        // Fetch previous transactions (inputs)
+        std::map<COutPoint, Coin> coins;
+        for (const CTxIn& txin : mtx.vin) {
+            coins[txin.prevout]; // Create empty map entry keyed by prevout.
+        }
+        pwallet->chain().findCoins(coins);
```
Based on the command description, why this is needed?

Shouldn't `simulaterawtransaction` only care about how the wallet would behave if a certain transaction is relayed to the network?.

In other words, doesn't seems to be ok to fetch inputs that are not owned/watched by this wallet as them don't mean anything for the wallet.

We could replace the input spendability check that is below using `wallet.IsSpent(outpoint)`.

kallewoof replied:
I'm not sure what you're proposing. Mind writing a patch that describes your suggestion?

furszy replied:
sure, it's merely about removing the coins fetching from chain and using the wallet information to answer the spendability question.

furszy replied:
Note: haven't tested it but check it out:
```
diff --git a/src/wallet/rpc/wallet.cpp b/src/wallet/rpc/wallet.cpp
--- a/src/wallet/rpc/wallet.cpp	(revision b557f53eac39b92bfc4922729e94087556de749a)
+++ b/src/wallet/rpc/wallet.cpp	(date 1659372880572)
@@ -654,13 +654,6 @@
             throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Transaction hex string decoding failure.");
         }
 
-        // Fetch previous transactions (inputs)
-        std::map<COutPoint, Coin> coins;
-        for (const CTxIn& txin : mtx.vin) {
-            coins[txin.prevout]; // Create empty map entry keyed by prevout.
-        }
-        pwallet->chain().findCoins(coins);
-
         // Fetch debit; we are *spending* these; if the transaction is signed and
         // broadcast, we will lose everything in these
         for (const auto& txin : mtx.vin) {
@@ -668,19 +661,21 @@
             if (spent.count(op)) {
                 throw JSONRPCError(RPC_INVALID_PARAMETER, "Transaction(s) are spending the same output more than once");
             }
+            spent.insert(op); // mark it as spent
             if (new_utxos.count(op)) {
                 changes -= new_utxos.at(op);
                 new_utxos.erase(op);
             } else {
-                if (!coins.count(op)) {
-                    throw JSONRPCError(RPC_INVALID_PARAMETER, "One or more transaction inputs are missing");
-                }
-                if (coins.at(op).IsSpent()) {
+                const auto* input_wtx = pwallet->GetWalletTx(txin.prevout.hash);
+                if (!input_wtx) continue;
+                if (input_wtx->tx->vout.size() >= txin.prevout.n) continue; // todo: this should throw an error
+
+                if (!(pwallet->IsMine(input_wtx->tx->vout[txin.prevout.n]) & filter)) continue;
+                if (pwallet->IsSpent(txin.prevout)) {
                     throw JSONRPCError(RPC_INVALID_PARAMETER, "One or more transaction inputs have been spent already");
                 }
                 changes -= pwallet->GetDebit(txin, filter);
             }
-            spent.insert(op);
         }
 
         // Iterate over outputs; we are *receiving* these, if the wallet considers
```

As said in the comment, it's about only using the wallet information to check the transaction and not the chain information. Mainly because, in this command, we only care about the wallet state.

kallewoof replied:
I'm not sure that's an improvement. The current implementation seems to be a bit more straightforward, even if it ends up pulling in a few extra inputs. If the difference in performance is significant it might be worth it though, but I'm dubious. :)

w0xlt commented on test/functional/wallet_simulaterawtx.py:
```diff
@@ -0,0 +1,129 @@
+        assert_raises_rpc_error(-8, "One or more transaction inputs have been spent already", w0.simulaterawtransaction, [tx3])
+        assert_raises_rpc_error(-8, "One or more transaction inputs have been spent already", w1.simulaterawtransaction, [tx3])
+        assert_raises_rpc_error(-8, "One or more transaction inputs have been spent already", w0.simulaterawtransaction, [tx4])
+        assert_raises_rpc_error(-8, "One or more transaction inputs have been spent already", w1.simulaterawtransaction, [tx4])
```
Should these 4 cases raise the exception `One or more transaction inputs are missing` ?

Anyway, this exception is not being tested.

kallewoof replied:
You're right. I don't think that exception will ever be triggered, because `coins` is always populated with all the outputs in `mtx.vin` in the earlier block. I think the best approach is to simply reword this exception to say "spent or missing" and drop the other one.

w0xlt commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -590,6 +590,120 @@ static RPCHelpMan upgradewallet()
+        // Fetch debit; we are *spending* these; if the transaction is signed and
+        // broadcast, we will lose everything in these
+        for (const auto& txin : mtx.vin) {
+            auto& op = txin.prevout;
```
nit: perhaps the `op` variable can be renamed to `outpoint` for the sake of clarity.

kallewoof replied:
Sure

w0xlt reviewed (approved):
tACK https://github.com/bitcoin/bitcoin/pull/22751/commits/b557f53eac39b92bfc4922729e94087556de749a

But as mentioned below, I think a test coverage for the exception `One or more transaction inputs are missing`  can be added.

achow101 commented:
re-ACK ced8a3e77d0dde428360f2250576582b624c013c

jonatack commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -590,6 +590,117 @@ static RPCHelpMan upgradewallet()
+            },
+            {"options", RPCArg::Type::OBJ_USER_KEYS, RPCArg::Optional::OMITTED_NAMED_ARG, "Options",
+                {
+                    {"include_watchonly", RPCArg::Type::BOOL, RPCArg::DefaultHint{"true for watch-only wallets, otherwise false"}, "Whether to include watch-only addresses (see 'importaddress')"},
```
https://github.com/bitcoin/bitcoin/commit/ced8a3e77d0dde428360f2250576582b624c013c
```suggestion
                    {"include_watchonly", RPCArg::Type::BOOL, RPCArg::DefaultHint{"true for watch-only wallets, otherwise false"}, "Whether to include watch-only addresses (see RPC importaddress)"},
```

jonatack commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -590,6 +590,117 @@ static RPCHelpMan upgradewallet()
+{
+    std::shared_ptr<const CWallet> const wallet = GetWalletForJSONRPCRequest(request);
+    if (!wallet) return NullUniValue;
+    const CWallet* pwallet = wallet.get();
```
https://github.com/bitcoin/bitcoin/commit/ced8a3e77d0dde428360f2250576582b624c013c my understanding was that a reference is preferred? (see `RPCHelpMan getbalances` for an example)
```suggestion
    const CWallet& wallet = *wallet;
```

jonatack reviewed (commented):
Light ACK ced8a3e77d0dde428360f2250576582b624c013c light review, rebased to current master, debug build, checked RPC help and new test

A few suggestions.

kallewoof commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -590,6 +590,117 @@ static RPCHelpMan upgradewallet()
+        // everything in these
+        // Also populate new_utxos in case these are spent in later transactions
+
+        const auto hash = mtx.GetHash();
```
Does that differ from just doing `const auto& hash = `? I assumed it wouldn't, but you're explicitly switching to `uint256` so now I'm unsure.

jonatack replied:
No difference between `auto` and `uint256`, just highlighting the return type to clarify why it is memoized by const ref (in the second version) rather than copied (as `auto` isn't really shorter or simpler in this case than just stating the type).

kallewoof commented:
Applied suggestions by @jonatack. Thanks for the feedback!

jonatack commented:
ACK db10cf8ae36693cb4d3ed1b47b84709cf9c0d849

<details><summary>A few minor ideas if you retouch (clang-tidy named args format, a const ref, a test)</summary><p>

```diff
--- a/src/wallet/rpc/wallet.cpp
+++ b/src/wallet/rpc/wallet.cpp
@@ -628,12 +628,12 @@ RPCHelpMan simulaterawtransaction()
 
     UniValue include_watchonly(UniValue::VNULL);
     if (request.params[1].isObject()) {
-        UniValue options = request.params[1];
+        const UniValue& options = request.params[1];
         RPCTypeCheckObj(options,
             {
                 {"include_watchonly", UniValueType(UniValue::VBOOL)},
             },
-            true, true);
+            /*fAllowNull=*/true, /*fStrict=*/true);
 
         include_watchonly = options["include_watchonly"];
     }
@@ -650,7 +650,7 @@ RPCHelpMan simulaterawtransaction()
 
     for (size_t i = 0; i < txs.size(); ++i) {
         CMutableTransaction mtx;
-        if (!DecodeHexTx(mtx, txs[i].get_str(), /* try_no_witness */ true, /* try_witness */ true)) {
+        if (!DecodeHexTx(mtx, txs[i].get_str(), /*try_no_witness=*/true, /*try_witness=*/true)) {
             throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Transaction hex string decoding failure.");
         }
 
@@ -685,7 +685,7 @@ RPCHelpMan simulaterawtransaction()
         // everything in these
         // Also populate new_utxos in case these are spent in later transactions
 
-        const auto& hash = mtx.GetHash(); // auto isn't shorter or simpler here so maybe be explicit about the type 
+        const uint256& hash = mtx.GetHash();
         for (size_t i = 0; i < mtx.vout.size(); ++i) {
             const auto& txout = mtx.vout[i];

--- a/test/functional/wallet_simulaterawtx.py
+++ b/test/functional/wallet_simulaterawtx.py
@@ -50,6 +50,9 @@ class SimulateTxTest(BitcoinTestFramework):
         tx1 = node.createrawtransaction([], [{address1: 5.0}])
         tx2 = node.createrawtransaction([], [{address2: 10.0}])
 
+        self.log.info("Test simulaterawtransaction result if no txns provided")  # drop this line if no other test logging
+        assert_equal(w0.simulaterawtransaction([]), {'balance_change': Decimal('0E-8')})
+
         # w0 should be unaffected, w2 should see +5 for tx1
         assert_equal(w0.simulaterawtransaction([tx1])["balance_change"], 0.0)
```
</p></details>

achow101 commented:
re-ACK db10cf8ae36693cb4d3ed1b47b84709cf9c0d849

Matchjuan859 commented on test/functional/wallet_simulaterawtx.py:
```diff
@@ -0,0 +1,129 @@
+        assert_approx(w2.simulaterawtransaction([tx2])["balance_change"], 0)
+
+        # w0-w2 error due to tx1 already being mined
+        assert_raises_rpc_error(-8, "One or more transaction inputs are missing or have been spent already", w0.simulaterawtransaction, [tx1, tx2])
```
test_

MarcoFalke commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -590,6 +590,117 @@ static RPCHelpMan upgradewallet()
+    return RPCHelpMan{"simulaterawtransaction",
+        "\nCalculate the balance change resulting in the signing and broadcasting of the given transaction(s).\n",
+        {
+            {"rawtxs", RPCArg::Type::ARR, RPCArg::Optional::OMITTED_NAMED_ARG, "An array of hex strings of raw transactions.\n",
```
Pretty sure this is not optional (should be `NO`, instead of `OMITTED_NAMED_ARG`)

MarcoFalke replied:
Also, the newline seems spurious?

MarcoFalke commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -590,6 +590,117 @@ static RPCHelpMan upgradewallet()
+RPCHelpMan simulaterawtransaction()
+{
+    return RPCHelpMan{"simulaterawtransaction",
+        "\nCalculate the balance change resulting in the signing and broadcasting of the given transaction(s).\n",
```
style nit: `\n` are not needed, as they don't change the help output

MarcoFalke commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -590,6 +590,117 @@ static RPCHelpMan upgradewallet()
+
+    for (size_t i = 0; i < txs.size(); ++i) {
+        CMutableTransaction mtx;
+        if (!DecodeHexTx(mtx, txs[i].get_str(), /* try_no_witness */ true, /* try_witness */ true)) {
```
Why are you trying no witness? It is impossible for a valid transaction to fail witness deserialization. See also `git grep "Make sure the tx has at least one input"`

MarcoFalke replied:
I see that the tests are relying on no-inputs txs, but I wonder if there is a use case to run this on transactions that can never be valid on the network, even if properly signed.

MarcoFalke commented on test/functional/wallet_simulaterawtx.py:
```diff
@@ -0,0 +1,129 @@
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def setup_network(self, split=False):
```
What is `split`?

MarcoFalke replied:
Why is this function needed in the first place?

MarcoFalke commented on test/functional/wallet_simulaterawtx.py:
```diff
@@ -0,0 +1,129 @@
+    def run_test(self):
+        node = self.nodes[0]
+
+        self.generate(node, 1, sync_fun=self.no_op) # Leave IBD
```
No need to skip the sync_fun. With one node, it will be a no-op anyway

MarcoFalke commented on test/functional/wallet_simulaterawtx.py:
```diff
@@ -0,0 +1,129 @@
+
+        # send tx1 to avoid reusing same UTXO below
+        node.sendrawtransaction(w0.signrawtransactionwithwallet(tx1)["hex"])
+        self.generate(node, 1, sync_fun=self.no_op) # Confirm tx to trigger error below
```
Same

MarcoFalke commented on test/functional/wallet_simulaterawtx.py:
```diff
@@ -0,0 +1,129 @@
+        # send tx1 to avoid reusing same UTXO below
+        node.sendrawtransaction(w0.signrawtransactionwithwallet(tx1)["hex"])
+        self.generate(node, 1, sync_fun=self.no_op) # Confirm tx to trigger error below
+        self.sync_all()
```
Can be removed
